Protection Icons Overhaul

Myszax opened the pull request:
I've extended `ShowTargetProtection` functionality.
Split `ShowTargetProtection` into two separate settings `ShowTargetProtectionNoFight` and `ShowTargetProtectionInFight`.
Why? When I am playing Gothic I've found that current settings does not me freedom to choose what I want to see on my screen.
For example: When I wandering around the game world talking with NPCs etc. I'd rather see their `All` protection info or see the ones that matters `AllButZeros`. In fight mode I want to be focused on `CurrentWeapon` only.
Players now will able to choose what fits them :)

System looks more consistent now. When NPC has `NPC_FLAG_IMMORTAL` there is no point in showing all icons, just `Shield` in gray color. When target is immune to some damage type then corresponding icon and value `MAX` is shown.

There is another option to choose icon types. "New" ones are the same as `DamagePopup` icons as suggested in #40 . Second option is to use `Shield` icons those that have been used so far.

Last but not least. System is now checking for all equipped weapon/spells damage types. Not only for `GetTopDmgIndex`.
For example Uriziel in G1 has magic and fire damage type. So for this weapon and `CurrentWeapon` setting both fire and magic protection icons will be shown.

Myszax committed:
Split `ShowTargetProtection` into two separate settings`ShowTargetProtectionNoFight` and `ShowTargetProtectionInFight`

Myszax committed:
Changed icon protecions from shields to damage popup icons

Myszax committed:
Note about immunities added to README.md

Myszax committed:
Added `TargetProtectionIconStyle` option to choose between old and new style

Myszax committed:
Changed italic style for protection icons decription

Myszax committed:
Gray color for protection icon is reserved only for `NPC_FLAG_IMMORTAL`. Value `MAX` is shown when target has immunity, but color stays same

Myszax committed:
Now it's checking all damage types instead of `GetTopDmgIndex`

Myszax commented:
I am wondering about `oEDamageIndex_Barrier`. Is it even considerable to show barrier damage type protection? I've never seen something with `oEDamageType_Barrier`.
Kinda same applies to `oEDamageIndex_Fall`. But I left it as it was originally.

Myszax commented:
Another thing came into my mind. Idk if it would be an overkill xD
Add fifth option for `ShowTargetProtectionNoFight`/`ShowTargetProtectionInFight` called `Custom` that player would be able to set custom order of icons and choose exactly those he wants to see.

Myszax commented:
Don't know if `spell->spellType == 2` should be checked before `MarkSpellDamage`.
There are 3 constants in G1/G2 scrips:
```d
const int SPELL_GOOD = 0;
const int SPELL_NEUTRAL = 1;
const int SPELL_BAD = 2;
```
0 is never used, 1 is used for no charm spells like (heal/light), 2 is used for all kind of damaging/transformation.
Could modification change this behavior?

Franisz commented:
> Seems like engine (at least G2A) is transforming `oETypeDamage` done by distance weapons to `oEDamageType_Point`. No matter if `ITM_CAT_FF` or `ITEM_KAT_MUN` (in current implementation I do not even check for amunition) has `damageType` different than `DAM_POINT` it's still converted. Tested on magic crossbow. Is there a point of checking damage values for distance weapons? Or just set bit for `oEDamageType_Point` and go on.

IIRC some mods or Ninja FreeAim allows for dealing different types of damage. I don't think there would be any universal solution for this so maybe it would be good to check for actual damage type dealt at runtime and flag whether dmg matches the weapon or the arrow or the fixed point damage type and save the flag to ini. On mismatch overwrite the flag.

Myszax committed:
Changed immune text value

Myszax committed:
Refactored `GetDamageIndexes`

Myszax committed:
Added null check for active spell in `BuildFightModeDamage` to avoid crashing

Myszax committed:
Mark munition damage type instead of distance weapon if possible

Myszax committed:
Distance weapon damage type marking depends on `OnDamage_Hit` hook

Myszax committed:
Added support for saving distance damage type to ini
Refactoring

Myszax committed:
Changed `MarkSpellDamage` to `MarkIntDamageType` and parameter

Franisz commented on zUtilities/FocusStatusBar.h:
```diff
@@ -9,8 +9,16 @@ namespace GOTHIC_ENGINE {
+		Disabled = 0,
+		CurrentWeapon = 1,
+		AllButZeros = 2,
+		All = 3
```
I think here should be only, `Disabled`, `CurrentWeapon`, `All` and there should be separate option for excluding from "All". We should be able to exclude Zeros, Fall, Fly and any combination of those. It should not affect the display of CurrentWeapon option.

Myszax replied:
Okay.
What about difference between fight and no fight mode? Should be unified like `TargetProtectionIconStyle` to not double options? I think so.
Then it will be 3 new options like `ProtectionIconHideZeros`, `ProtectionIconHideFall`, `ProtectionIconHideFly` and it will be applicable only when mode is set to `All`?
It would be even better to add mask configuration, but it will make people confused how to set it up.
It could be expanded to 8 booleans (7 for damage types and 1 for hide zeros) - simpler and more elastic for players.
Please elaborate.

Franisz replied:
I think those 3 separate options that apply for both in fight and out of combat is fine. Mask would be ok as well, but I don't really see why someone would want to set it up completely different for combat/out of combat, so simpler is better here i think.

Franisz commented on zUtilities/PlayerStatus.h:
```diff
@@ -26,15 +27,14 @@ namespace GOTHIC_ENGINE {
-      ShowCurrWeapProtOnly = showTargetProtectionValue == 1;
-      ShowProtOnlyInFight = zoptions->ReadBool(PLUGIN_NAME, "ShowProtOnlyInFight", true);
-      ShowProtAllDamageTypes = zoptions->ReadBool(PLUGIN_NAME, "ShowProtAllDamageTypes", false);
+      ShowTargetProtectionNoFight = zoptions->ReadInt(PLUGIN_NAME, "ShowTargetProtectionNoFight", TargetProtectionMode::AllButZeros);
```
This should be disabled by default.

Franisz commented on zUtilities/NpcHelper.cpp:
```diff
@@ -96,4 +172,94 @@ namespace GOTHIC_ENGINE {
+			DamageMask tmp{};
+			MarkIntDamageType(desc.enuModeDamage, tmp);
+			Options::DistanceWeaponDamageType = tmp;
+			zoptions->WriteInt(PLUGIN_SECTION_TEMP, "DistanceWeaponDamageType", desc.enuModeDamage, 0);
```
Don't create separate section and add trivia for this entry.

Franisz commented:
<img width="205" height="146" alt="01-02-2026_12-07-39" src="https://github.com/user-attachments/assets/5dbcd8ec-8a8b-4691-8c04-2a00ea71a7d1" />

With using dmg icons now spacing looks bad in case of some icons. Maybe just increase the spacing a bit.

Myszax committed:
Added little spacing between protection icons

Myszax commented:
> <img alt="01-02-2026_12-07-39" width="205" height="146" src="https://private-user-images.githubusercontent.com/47374092/543413171-5dbcd8ec-8a8b-4691-8c04-2a00ea71a7d1.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAxMzkwNDQsIm5iZiI6MTc3MDEzODc0NCwicGF0aCI6Ii80NzM3NDA5Mi81NDM0MTMxNzEtNWRiY2Q4ZWMtOGE4Yi00NjkxLThjMDQtMmEwMGVhNzFhN2QxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjAzVDE3MTIyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZkMDZlZjVhYzkzMmVkZjc2YTBiMWQ4YjU3YjhhOTJmNjVmYmQ2MTIxZWI5NzkyMTg1OTJmOGUxNjM0Nzc4ZDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.l3h8pVFb6YMjiQVym8CJ7BwFpY-v03hOidKvztHsD4k">
> 
> With using dmg icons now spacing looks bad in case of some icons. Maybe just increase the spacing a bit.

Sure it looks odd.
I've added same spacing for both versions. It looks kinda long :P
<img width="498" height="129" alt="obraz" src="https://github.com/user-attachments/assets/3a79c0d6-ef1c-4ad3-9d8b-ed0989ffa550" />
<img width="504" height="119" alt="obraz" src="https://github.com/user-attachments/assets/15b37fb1-3aa5-40b6-a536-bb48ff6e279b" />
Idk, maybe i should add it (or increase further) only to dmg icons, but it looks better in both versions

Franisz commented:
> > <img alt="01-02-2026_12-07-39" width="205" height="146" src="https://private-user-images.githubusercontent.com/47374092/543413171-5dbcd8ec-8a8b-4691-8c04-2a00ea71a7d1.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAxMzkwNDQsIm5iZiI6MTc3MDEzODc0NCwicGF0aCI6Ii80NzM3NDA5Mi81NDM0MTMxNzEtNWRiY2Q4ZWMtOGE4Yi00NjkxLThjMDQtMmEwMGVhNzFhN2QxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjAzVDE3MTIyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZkMDZlZjVhYzkzMmVkZjc2YTBiMWQ4YjU3YjhhOTJmNjVmYmQ2MTIxZWI5NzkyMTg1OTJmOGUxNjM0Nzc4ZDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.l3h8pVFb6YMjiQVym8CJ7BwFpY-v03hOidKvztHsD4k">
> > With using dmg icons now spacing looks bad in case of some icons. Maybe just increase the spacing a bit.
> 
> Sure it looks odd. I've added same spacing for both versions. It looks kinda long :P <img alt="obraz" width="498" height="129" src="https://private-user-images.githubusercontent.com/6831024/544495274-3a79c0d6-ef1c-4ad3-9d8b-ed0989ffa550.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAyMDM4NTYsIm5iZiI6MTc3MDIwMzU1NiwicGF0aCI6Ii82ODMxMDI0LzU0NDQ5NTI3NC0zYTc5YzBkNi1lZjFjLTRhZDMtOWQ4Yi1lZDA5ODlmZmE1NTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDIwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAyMDRUMTExMjM2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjU1ZDMwOGM4MDg0MzkzOTY4N2UxOWMyY2ZjZjdlMTBjZjM3NzdiMTU0ZDFiYzEyYjQzM2ZkMjhlMTRlYmUyOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.kbH9NQ-geEGZGiL7ITtG5Hhcz3OYqz4c4bq1F5WZ02Y"> <img alt="obraz" width="504" height="119" src="https://private-user-images.githubusercontent.com/6831024/544496924-15b37fb1-3aa5-40b6-a536-bb48ff6e279b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAyMDM4NTYsIm5iZiI6MTc3MDIwMzU1NiwicGF0aCI6Ii82ODMxMDI0LzU0NDQ5NjkyNC0xNWIzN2ZiMS0zYWE1LTQwYjYtYTUzNi1iYjQ4ZmY2ZTI3OWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDIwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAyMDRUMTExMjM2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjExNTk5YTcwZWU0MTg3NzBiYjAzZGIzMzUxYjM3NDY3ZDgyY2RkMzU5OWM1MWY5MmNiYWZhMDVjYjk0ZmI3YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.a7LtsT3GYwnKNYTx4qCudVEuJ3irFBQy0u1Vv_Joc-8"> Idk, maybe i should add it (or increase further) only to dmg icons, but it looks better in both versions

Spacing in both versions is fine. It's inevitable to further clutter the screen the more protections we want to display, that's why the default in plugin is single value in combat only.

Initially I also thought about listing it in column next to the npc, but dunno if that would be better.
<img width="547" height="685" alt="04-02-2026_12-48-16" src="https://github.com/user-attachments/assets/465a323a-9938-42eb-9a7e-ef3b395084d9" />

Myszax committed:
Removed `ZUTILITIES_TEMP` section
Added trivia for `DistanceWeaponDamageType`

Myszax committed:
Fixed writing `DistanceWeaponDamageType` value to ini

Myszax commented:
Quick preview
<img width="534" height="392" alt="img" src="https://github.com/user-attachments/assets/06b4050b-956a-4293-8d52-36f80ee36566" />
I don't know what to think about it 😃
Emm this could be added like an another option, but idk.

Franisz commented:
I think its fine as an option. It's definitely more readable especially for mods that tend to have some absurd prot values like `99999` in Odyssee for example.

Myszax committed:
Added missing `DistanceWeaponDamageType` to auto generated ini

Myszax committed:
Added new option `TargetProtectionIconPosition`

Myszax committed:
Refactor for rendering methods

Myszax committed:
Added option to hide some protection icons

Myszax committed:
Removed `AllButZeros` from `TargetProtectionMode`

Myszax committed:
merge

Myszax committed:
Added `BuildRenderData`

Myszax committed:
Added `DamageMaskHelper`
Removed `NpcHelper`

Myszax committed:
Removed `GetProtectionStatusesVisibleCount`

Myszax committed:
Added `ProtectionRenderMode` and `ProtectionContext`

Myszax committed:
Cleanup dummy arguments from StatusBar

Myszax committed:
Removed `const_cast` from `RenderProtection`. It uses reference to vector now

Myszax committed:
Refactor + added new margin method

Myszax committed:
Fixed `RenderProtectionIconsRight` spacing between icons

Myszax committed:
Divide into model+layout+render

Myszax committed:
Removed per-frame vector allocations

Myszax committed:
Render methods are using layout as parameter from now on

Myszax committed:
`protectionModel` is now cached
added `isInFightMode` and `TargetProtectionMode` to it

Myszax committed:
rename fields in `ProtectionModel`

Myszax committed:
Added cache for icons (partly)

Myszax committed:
Refactored `BuildProtectionLayout`
Cache for `protView->FontY()`
Removed unused `GetProtSize`, `GetProtStartX`, `GetProtStartY`

Myszax committed:
`ProtectionLayout` changed `size` to `iconSize
Removed unused `IsShowTargetProtectionDisabled`

Myszax committed:
Switched from vector to normal array to squeeze more optimization

Myszax committed:
Reorder variables and methods in files

Myszax committed:
Revert .sln and .gitignore changes from 98a4409

Myszax commented:
I rewrote the entire `FocusStatusBar` class.
### Main changes

- Split logic into **Model + Layout + Render**
- Render now operates on precomputed layout
- Removed most per-frame allocations
- Switched from `vector` to array to avoid per-frame allocations
- Added caching for `ProtectionModel` and selected layout values
- Added `ProtectionRenderMode` and `ProtectionContext`
- Removed unused methods and legacy helpers
- General cleanup and method/variable reordering

### Impact

- Reduced memory allocations
- Cleaner and more maintainable architecture
- Easier to extend and debug
- No intended gameplay changes

### Performance notes

Benchmarking was done in a simple way:  
- `std::chrono::high_resolution_clock::now()` at the start and end of `NeedAdjustPosition`  
- Duration calculated as `(end - start).count() / 1000.0f` in microseconds  
- Logged every 60 ticks; test subject had 6 icons

Observations:
- Performance gain is minor for a small number of icons  
- With 1 icon, render time is unchanged  
- At 60 FPS, ~10 microseconds faster  
- At 144 FPS, average times are almost the same  
- Gains become more noticeable with more icons

### Class structure improvements

- Separation into **Model, Layout, and Render** layers makes the class more modular  
- Easier to maintain and extend without touching unrelated code  
- Future rendering modes can be added with minimal impact on existing logic 

---
`iconCache` is trivial. In `ProtectionIconRenderData ` copies strings and then `IconInfo` does it too. I tried to avoid as much allocations as I can :)
I am wondering if cost of switching from `vector` to array is valuable. Vector is less bug prone and easier to maintain in code.
What about creating zCView every frame `protView = new zCView(0, 0, 8192, 8192);`?
Shouldn't be created once and then eventually `screen->RemoveItem`?
`iconCache[oEDamageIndex_Barrier] = "DMGICON_BARRIER";` don't even have own icon and probably will not be never need but *if yes* then I propose this icon :D https://game-icons.net/1x1/delapouite/lightning-dome.html

Myszax committed:
Added filter for `FocusStatusBar`

Myszax commented:
In G1 setting screen looks like
<img width="901" height="679" alt="obraz" src="https://github.com/user-attachments/assets/b4c20d27-da22-4f26-a62f-7587f27209f7" />

Also more options for changing spacing between icons can be added.

Myszax committed:
Moved `FocusStatusBar` filter into `PlayerStatus`

Myszax committed:
Merge branch 'master' into dev

Myszax committed:
Continuation of upgrading to 1.0m

Myszax committed:
Merge branch 'Franisz:master' into master

Myszax committed:
Revert "Continuation of upgrading to 1.0m"

This reverts commit 320abfb158d915c6b8e93e9743e237da9617ca7c.

Myszax committed:
Merged master

Myszax committed:
Fixed indents and space style
